nixos/modules/services/networking/adguardhome.nix

```diff
@@ -19,7 +19,9 @@ let
 
   configFile = pkgs.writeTextFile {
     name = "AdGuardHome.yaml";
-    text = builtins.toJSON (recursiveUpdate cfg.settings baseConfig);
+    text = builtins.toJSON (recursiveUpdate ({
+      schema_version = pkgs.adguardhome.passthru.schema_version;
```

Comment:
Add this to `config.services.adguardhome.settings` instead?
```
config = mkIf cfg.enable {
  services.adguardhome.settings = pkgs.adguardhome.passthru.schema_version;
};
```

Some advanced users may want to override this with `lib.mkForce` and pin to a specific schema? If I'm completely off feel free to ignore this comment.

Also, is the `passthru` in `passthru.schema_version` actually required there? :thinking:

Reply:
Regarding overwriting `schema_version` in a config: Note that I've added an explicit `schema_version = 0` to the tests, so that is working properly (for example, `dns.bind_host` was renamed to `dns.bind_hosts` in [v0.106.0](https://github.com/AdguardTeam/AdGuardHome/wiki/Configuration#configuration-file)).

Reply:
Is settings not a proper freeform module?

Reply:
It is nearly a proper freeform option, but there are two externally defined options (`host` and `port`). As a result, the sequence of attrset-updates is:

 1. Default `schema_version`
 2. `baseConfig` (which uses the options `host` and `port` external to `settings`)
 3. Other changes due to `settings`.

But now that I think about it: It should be possible to simply drop `baseConfig`. That option is more or less a relic of `host` and `port` used in [the arguments of the service](https://github.com/NixOS/nixpkgs/commit/24b8c37281cf339d1fab1098dd700674ca3cb9e7#diff-ccdc8494496fa82e2c07274c6267d2339b7f5223c89a36762cd1c59e2deec32aL13-L14) instead of defining them in the settings.

That would break backwards compatibility due to removing two options (which is probably the right thing to do here).

Reply:
I have translated `settings` to a proper freeform module, removing the need for any `recursiveUpdate` or similar composing technique.